### PR TITLE
rewrite term in ncurses

### PR DESCRIPTION
**Pull request by BernardoDeLaPlaz**
Integrating with master will require either:

(1) a breaching change
(2) a new pill (which I can generate)
(3) the following sequence of steps:  
- checkout arvo ncurses branch
- compile and run this version of urbit with some old pill
- dojo will appear screwed up. Each letter you type will cause the prompt to be echoed. Ignore that. Type "|mount %"
- cp  ../arvo/sys/zuse.hoon zod/home/sys/zuse.hoon 
- cp ../arvo/sys/vane/dill.hoon zod/home/sys/vane/dill.hoon 
- done

### CODE REVIEW

**Comment by ohAitch** [comment]
A new pill certainly sounds like a breaching change!

Could you expand on why this cannot e.g. stay in non-ncurses compatibility
mode until it receives one of the new motes? Obviously we'd want to take
that out in the long run, but I don't see why "each letter you type will
echo the prompt" type artifacts are difficult to avoid.

Also, how does this interact with the automated testing infrastructure,
does the output log look relatively sane? (We do eventually want to move
those to urb.py operation that will make this less of a concern, but iirc
we haven't yet)

File: include/c/motes.h [hunk 1]
```diff
@@ -96,7 +96,9 @@
 #   define c3__bnpd   c3_s4('b','n','p','d')
 #   define c3__bnps   c3_s4('b','n','p','s')
 #   define c3__bnsp   c3_s4('b','n','s','p')
+#   define c3__boat   c3_s4('b','o','a','t')
 #   define c3__boce   c3_s4('b','o','c','e')
+#   define c3__bog    c3_s3('b','o', 'g')  /* set background color in term */
```
Comment (wcable) [inline include/c/motes.h]: are the stray spaces on commented mote lines intentional?
relatedly, trying to document motes here seems very at odds with the style of the rest of the file: I think `motes.h` is meant to be a very mechanical "fully macro-expanded" minor C language extension, while semantics commentary should go in the relevant `term.c` case statements and hoon structure definitions

Comment (BernardoDeLaPlaz) [reply include/c/motes.h]: fixed

**Comment by BernardoDeLaPlaz** [comment]
> A new pill certainly sounds like a breaching change! Could you
> expand on why this cannot e.g. stay in non-ncurses compatibility
> mode until it receives one of the new motes?

The short answer is that this is absolutely something we can do.  I had never learned about breaches, coordination of arvo and urbit releases, etc. until after the project was feature complete / Anthony was told "it's done".

Because I didn't know that trying to avoid requiring synchronized changes was a goal, I didn't even think about it.

Also, there was a large amount of undocumented code in term.c that, after struggling with for a while, I just ripped out and rewrote.

The upshot is that  dill.hoon uses 5 or so blits in term.c, and those blits overlap with the syntactic names used before, but at least one of them has slightly different semantics.

If I'm told "do some more work to make it a non-breaching change", then I'm happy to do so. But that question impacts the schedule of other projects. Thoughts @vvisigoth ?

**Comment by BernardoDeLaPlaz** [comment]
>  Also, how does this interact with the automated testing infrastructure, does the output log look relatively sane? 

It seems to break it

https://travis-ci.org/urbit/urbit/builds/509915232?utm_source=github_status&utm_medium=notification

**Comment by ohAitch** [comment]
Ah, I suppose I should have just checked the github badge. Well if it fails
the tests, I will definitely contest "it's done", clearly it is necessary
to fix either it or the tests :P

**Comment by BernardoDeLaPlaz** [comment]
>  I will definitely contest "it's done", clearly it is necessary to fix either it or the tests :P

valid.

Both for my own immediate edification, and for the doc I'm writing, what is the correct way to run the tests in devel ?

**Comment by ohAitch** [comment]
hopefully in this case the `npm run -s test` command which was the step
that failed on the linked build?
https://github.com/urbit/urbit/blob/master/.travis.yml is going to be the
general entry point, wrt dependencies and such, everything except the npm
install is likely to already be on your machine as part of regular vere
development; I guess "cd .travis" is another important step.

**Comment by philipcmonk** [comment]
> (We do eventually want to move those to urb.py operation that will make this less of a concern, but iirc we haven't yet)

This is done on cc-release, but not on master.

> If I'm told "do some more work to make it a non-breaching change", then I'm happy to do so.

We're breaching regularly, I think this easily falls into the category of "just merge into 'next' and wait for the next breach".

**Comment by ohAitch** [comment]
That's fair, it seems like keeping compatibility temporarily might be
useful for the tests(and not incompatibly redefining what motes mean is
just polite), but probably just changing those is more straightforward. (It
does seem important for urbit to be able to run in daemon mode where it
produces logs without ANSI escapes everywhere, but w/e daemon mode is
broken in a number of other ways, and most of the hairy term code thrown
out is probably not necessary for just emitting lines verbatim)

I do maintain that "either: (1) a breach (2) a new pill or (3) here's how
to generate that pill" is somewhat confusing, given the plan is to do all
three :P

**Comment by Fang-** [comment]
Pinging for status/proper management of this and urbit/arvo#1115. Would love to see this get in soon! Happy to review myself if there's no other takers, but I bet there's people better suited than me.

**Comment by BernardoDeLaPlaz** [comment]
@Fang- When I checked it in, I thought it was good, but after more thought I think it might have a few problems.  I realized later that the "silly" << ... >> text that I removed was actually doing something (printing the current activity to screen, IIRC).  I'd love to have someone not just code review it, but actually poke at it and use it for 5 minutes.  Maybe there are a few bugs I need to fig before it gets merged in?  Maybe we merge it in ASAP and then file bugs against it?  Either way, I'd also love to have this no longer hanging over my head .

cc @vvisigoth

**Comment by Fang-** [comment]
If you can set me up with "how to launch this" instructions (though maybe that's just what you wrote in the OP?) then I might poke at this a bit over the weekend to play around and see if there's any show stoppers/compile a list of obvious issues.

**Comment by BernardoDeLaPlaz** [comment]
Oh, gosh, now I remember that it requires a matched set of changes in vere and in arvo, and I so the idea is that we want to save it for the next breaching change.

but, still , it'd be great to play with it ahead of time to compile that list, as you say

yes, see OP for "how to" ; ping me with any questions

**Comment by ohAitch** [comment]
> - Can't scroll up, so no way to fully see any prints that are taller than the screen. Is there an easy fix for this, or do we need to capture scroll events and maintain scrollback state?

It seems a lot better to figure out how to append lines to the native terminal scrollback log properly, "delete them until retrieved by a scroll event" isn't super useful if urbit crashes or ssh disconnects or w/e. If the fix is easy I think it's "don't use the alternate terminal buffer"?

**Comment by jtobin** [comment]
Per my commentary in urbit/arvo#1115: this one appears to require a bit more work before it's suitable to be pulled in, and it's not clear when that will occur, so I'll close it for the time being.  Feel free to reopen when you reckon it might be good to go.

**Pull request closed without merging**